src/main/java/org/apache/sysml/hops/AggBinaryOp.java

```diff
@@ -344,11 +344,30 @@ protected double computeOutputMemEstimate( long dim1, long dim2, long nnz )
 	protected double computeIntermediateMemEstimate( long dim1, long dim2, long nnz )
 	{
 		double ret = 0;
-		
+
+		if (DMLScript.USE_ACCELERATOR) {
+			// In GPU Mode, these are the intermediate estimates for matmult:
+			// Both dense or both sparse - no intermediate
+			// One sparse, one dense - (dim1 * dim2) OR (dim1 * dim2 + input in sparse format)
+			// In the worst case, an intermediate block other other than a potential temporary sparse-dense block
+			// can be created.
+			Hop in1 = _input.get(0);
+			Hop in2 = _input.get(1);
+			boolean in1Sparse = MatrixBlock.evalSparseFormatInMemory(in1.getDim1(), in1.getDim2(), in1.getNnz());
+			boolean in2Sparse = MatrixBlock.evalSparseFormatInMemory(in2.getDim1(), in2.getDim2(), in2.getNnz());
```

Comment:
Use the OptimizerUtils estimation functions in order to guard against unknown nnz - theses methods here assume known nnz and compute garbage otherwise. Note that you need to consider the worst-case though, meaning if you don't know the sparsity you need to consider the largest possible sparse matrix (see sparsity turn point).